codex/contracts/market.nim

```diff
@@ -80,7 +96,7 @@ method subscribeRequests(market: OnChainMarket,
                          callback: OnRequest):
                         Future[MarketSubscription] {.async.} =
   proc onEvent(event: StorageRequested) {.upraises:[].} =
-    callback(event.requestId, event.ask)
+    asyncSpawn callback(event.requestId, event.ask)
```

Comment:
I would avoid `asyncSpawn` on callbacks. We cannot yet specify that async callbacks are not allowed to raise exceptions, and if they do happen to raise an exception, then the node crashes.

Ideally we'd keep the callbacks non-async. When we really to need to call async code in a callback, then we can do the `asyncSpawn` inside the callback; it's then easier to check that the async code doesn't raise any exceptions.

Reply:
why not just `await` the callback?

Reply:
> why not just `await` the callback?

`onEvent` is not async. One option would be to allow async event subscriptions in ethers. Something like this: https://github.com/status-im/nim-ethers/pull/35

Reply:
I pushed a commit that supports this: https://github.com/status-im/nim-codex/pull/306/commits/c2a8cea91b117d298b44cf1c0e8ff3728ec22cc9.

Reply:
That's is one way of solving the problem, but the disadvantage is that it keeps spreading `{.async.}` further and further (even into the nim-ethers library). I'd like to suggest an alternative approach: make sure that the state machine can handle normal (sync?) event handlers while allowing async behavior in the states. I have an idea on how to do that and will work this out shortly, so that we can compare approaches.

Reply:
What are the disadvantages of spreading `{.async.}` further and into ethers (which already uses async handlers for other event, eg new blocks)?

Reply:
> make sure that the state machine can handle normal (sync?) event handlers while allowing async behavior in the states

Do you mean having states that support both `enter` and `enterAsync` for example?

I can agree with that, but I believe in this particular situation, the way the async state machine is setup in this PR, we need to support async callbacks. Here's my logic... It's handling an event subscription in a contract. With the suggested changes in the ethers PR, we have the option to use sync or async event handlers, depending on what is required. In this particular case, the event handler callback enters a new state asynchronously (hence the need for an async callback). The async state transition is required so that the current state can be cancelled, allowing new contract events to come in and change the state. Using an `asyncSpawn` in the callback would not work in this case as the state machine would not have a handle on the current state and thus would not be able to cancel it.  If everything was synchronous, the entire `enter` state transition will need to complete its run *before* any new contract event handlers can run (which may, in turn, transition the state machine to a new state). We already saw that this can be a problem if there's any kind of network or IO interactions in a state. 

I'm very curious to see the solution you are working on. Happy to jump on a call with you this evening to discuss futher 👍

Reply:
> What are the disadvantages of spreading `{.async.}` further and into ethers (which already uses async handlers for other event, eg new blocks)?

Async code adds complexity and is more difficult to reason about than sync code, so as a general rule I try to avoid it except where necessary. When we run into a boundary between async and sync code (as we do with the synchronous event handlers that need to perform an async cancelAndWait in the state machine),  we can either make the calling code async, or carefully design the boundary to support the use case. 

If we make the calling code async, this may require more code to become async, which requires more code to become async, etc. Until we run into a dependency that we do not control and therefore cannot make async. At which point we will be forced to design the sync / async boundary. I prefer to contain the async code, and design the sync / async boundary early on.

I'll try to code out an alternative to the state machine with a carefully designed sync/async boundary soon. Let's jump on that call, and discuss it :smile:

Reply:
Added a couple of commits that introduce the new state machine design from #344 into this PR, which allows for synchronous callbacks.